From f479fb7a0fe6ed57c7535f2fc831ebeb8dae2646 Mon Sep 17 00:00:00 2001
From: Brent Murphy <bmurphy@bcmcs.net>
Date: Tue, 28 Apr 2020 13:18:36 +1000
Subject: [PATCH] [cmake] addons - handle exact matches and regexes distinctly

ADDONS_TO_BUILD="game.libretro"
Exact match - Will match game.libretro and only game.libretro

ADDONS_TO_BUILD="game.libretro.mr"
No exact match - considered regex, will match game.libretro.mrboom

ADDONS_TO_BUILD="game.libretro.*"
No exact match - considered regex, will match game.libretro and all other game.libretro.*
---
 cmake/addons/CMakeLists.txt | 64 +++++++++++++++++++++++++++----------
 1 file changed, 48 insertions(+), 16 deletions(-)

diff --git a/cmake/addons/CMakeLists.txt b/cmake/addons/CMakeLists.txt
index 20f2b6df4fb9..1c2685abc346 100644
--- a/cmake/addons/CMakeLists.txt
+++ b/cmake/addons/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.5)
+cmake_minimum_required(VERSION 3.6)
 project(kodi-addons)
 
 list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR})
@@ -244,6 +244,7 @@ endif()
 # error either in ADDONS_TO_BUILD or in the directory configuration.
 set(SUPPORTED_ADDON_FOUND FALSE)
 
+# Addon exclusion list
 set(EXCLUDE_ADDONS "")
 foreach(addon ${ADDONS_TO_BUILD})
   set(FOUND_EXCLUSION "")
@@ -252,18 +253,43 @@ foreach(addon ${ADDONS_TO_BUILD})
     # Remove beginning - and then append to excluded addon list
     STRING(REGEX MATCH "-([^ ]*)"  excluded_addon ${FOUND_EXCLUSION})
     list(APPEND EXCLUDE_ADDONS ${CMAKE_MATCH_1})
-    message(STATUS "Exclusion found ${CMAKE_MATCH_1}")
   endif()
 endforeach()
 
+message(STATUS "Exclusion list: ${EXCLUDE_ADDONS}")
+
+# This block gets us the list of exact matches
+set(EXACT_MATCH_ADDON_LIST "")
+set(REGEX_ADDONS_TO_BUILD ${ADDONS_TO_BUILD})
+# remove platforms.txt entries from list - not needed
+# output - REGEX_ADDONS_TO_BUILD
+list(FILTER addons EXCLUDE REGEX platforms.txt)
+foreach(addon ${ADDONS_TO_BUILD})
+  set(FILTER_LIST ${addons})
+  # need to strip regex chars, or the filter regex will use
+  string(REPLACE "*" "" strippedregex ${addon})
+  list(FILTER FILTER_LIST INCLUDE REGEX ".*${strippedregex}[.]txt")
+  if(NOT ${FILTER_LIST})
+    list(APPEND EXACT_MATCH_ADDON_LIST ${addon})
+    # remove exact matches from addons_to_build
+    list(REMOVE_ITEM REGEX_ADDONS_TO_BUILD "${addon}")
+  endif()
+endforeach()
+
+message(STATUS "Exact Match list: ${EXACT_MATCH_ADDON_LIST}")
+message(STATUS "Regex list: ${REGEX_ADDONS_TO_BUILD}")
+
 foreach(addon ${addons})
-  if(NOT (addon MATCHES platforms.txt))
+#  if(NOT (addon MATCHES platforms.txt))
+    # Parses <addon>.txt - addonname repo branch
     file(STRINGS ${addon} def)
     string(REPLACE " " ";" def ${def})
+    # id = addonname
     list(GET def 0 id)
 
-    list(FIND ADDONS_TO_BUILD ${id} idx)
+    # search exclusion rules for regex match of id
     set(ADDON_EXCLUDE FALSE)
+    set(ADDON_FOUND FALSE)
     foreach(exclusion ${EXCLUDE_ADDONS})
       if(id MATCHES "${exclusion}")
         set(ADDON_EXCLUDE TRUE)
@@ -276,19 +302,25 @@ foreach(addon ${addons})
       continue()
     endif()
 
-    set(ADDON_FOUND FALSE)
-    # try to find a perfect match
-    if(idx GREATER -1 OR "${ADDONS_TO_BUILD}" STREQUAL "all")
+    if("${ADDONS_TO_BUILD}" STREQUAL "all")
       set(ADDON_FOUND TRUE)
-    # Maybe we have a regex
     else()
-      foreach(ADDONLISTITEM ${ADDONS_TO_BUILD})
-        if(id MATCHES "${ADDONLISTITEM}")
-          message(STATUS "Pattern ${ADDONLISTITEM} matches ${id}, building addon")
-          set(ADDON_FOUND TRUE)
-          break()
-        endif()
-      endforeach()
+      # search exact match - -1 if not found
+      list(FIND EXACT_MATCH_ADDON_LIST ${id} exactidx)
+      if(exactidx GREATER -1)
+        # exact match, so build
+        message(STATUS "Exact match ${id}, building addon")
+        set(ADDON_FOUND TRUE)
+      else()
+        # regex search
+        foreach(ADDONLISTITEM ${REGEX_ADDONS_TO_BUILD})
+          if(id MATCHES "${ADDONLISTITEM}")
+            message(STATUS "Pattern ${ADDONLISTITEM} matches ${id}, building addon")
+            set(ADDON_FOUND TRUE)
+            break()
+          endif()
+        endforeach()
+      endif()
     endif()
 
     if(ADDON_FOUND)
@@ -460,7 +492,7 @@ foreach(addon ${addons})
         add_custom_target(${id} COMMAND ${CMAKE_COMMAND} -E echo "IGNORED ${id} - not supported on ${CORE_SYSTEM_NAME}\n")
       endif()
     endif()
-  endif()
+#  endif()
 endforeach()
 message(STATUS "")
 
